core/arch/arm/mm/mobj.c

```diff
@@ -312,6 +308,291 @@ struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+	case 0:
+		p = mobj_reg_shm->pages[offst / SMALL_PAGE_SIZE] +
+			offst % SMALL_PAGE_SIZE;
+		break;
+	case SMALL_PAGE_SIZE:
+		p = mobj_reg_shm->pages[offst / SMALL_PAGE_SIZE];
+		break;
+	default:
+		return TEE_ERROR_GENERIC;
+
+	}
+	*pa = p;
+
+	return TEE_SUCCESS;
+}
+
+static void mobj_reg_shm_free(struct mobj *mobj)
+{
+	struct mobj_reg_shm *mobj_reg_shm = to_mobj_reg_shm(mobj);
+	uint32_t exceptions;
+
+	exceptions = cpu_spin_lock_xsave(&reg_shm_slist_lock);
+	SLIST_REMOVE(&reg_shm_list, mobj_reg_shm,
+		     mobj_reg_shm, next);
+	cpu_spin_unlock_xrestore(&reg_shm_slist_lock, exceptions);
+	free(mobj_reg_shm);
+}
+
+static TEE_Result mobj_reg_shm_get_cattr(struct mobj *mobj __unused,
+					 uint32_t *cattr)
+{
+	if (!cattr)
+		return TEE_ERROR_GENERIC;
+
+	*cattr = TEE_MATTR_CACHE_CACHED;
+
+	return TEE_SUCCESS;
+}
+
+static const struct mobj_ops mobj_reg_shm_ops __rodata_unpaged = {
+	.get_pa = mobj_reg_shm_get_pa,
+	.get_cattr = mobj_reg_shm_get_cattr,
+	.free = mobj_reg_shm_free,
+};
+
+static struct mobj_reg_shm *to_mobj_reg_shm(struct mobj *mobj)
+{
+	assert(mobj->ops == &mobj_reg_shm_ops);
+	return container_of(mobj, struct mobj_reg_shm, mobj);
+}
+
+struct mobj *mobj_reg_shm_alloc(paddr_t *pages, size_t num_pages,
+				paddr_t page_offset, uint64_t cookie)
+{
+	struct mobj_reg_shm *mobj_reg_shm;
+	size_t i;
+	unsigned int exceptions;
+
+	if (!num_pages)
+		return NULL;
+
+	for (i = 0; i < num_pages; i++) {
+		if (pages[i] & SMALL_PAGE_MASK)
+			return NULL;
+
+		/* Only Non-secure memory can be mapped there */
+		if (!core_pbuf_is(CORE_MEM_NON_SEC, pages[i], SMALL_PAGE_SIZE))
+			return NULL;
+	}
+
+	mobj_reg_shm = calloc(1, MOBJ_REG_SHM_SIZE(num_pages));
+	if (!mobj_reg_shm)
+		return NULL;
+
+	mobj_reg_shm->mobj.ops = &mobj_reg_shm_ops;
+	mobj_reg_shm->mobj.size =  num_pages * SMALL_PAGE_SIZE;
+	mobj_reg_shm->mobj.granule = SMALL_PAGE_SIZE;
+	mobj_reg_shm->cookie = cookie;
+	mobj_reg_shm->num_pages = num_pages;
+	mobj_reg_shm->page_offset = page_offset;
+	memcpy(mobj_reg_shm->pages, pages, sizeof(*pages) * num_pages);
+
+	exceptions = cpu_spin_lock_xsave(&reg_shm_slist_lock);
+	SLIST_INSERT_HEAD(&reg_shm_list, mobj_reg_shm, next);
+	cpu_spin_unlock_xrestore(&reg_shm_slist_lock, exceptions);
+
+	return &mobj_reg_shm->mobj;
+}
+
+struct mobj *mobj_reg_shm_find_by_cookie(uint64_t cookie)
+{
+	struct mobj_reg_shm *mobj_reg_shm;
+	unsigned int exceptions;
+
+	exceptions = cpu_spin_lock_xsave(&reg_shm_slist_lock);
+	SLIST_FOREACH(mobj_reg_shm, &reg_shm_list, next) {
+		if (mobj_reg_shm->cookie == cookie) {
+			cpu_spin_unlock_xrestore(&reg_shm_slist_lock,
+						 exceptions);
+			return &mobj_reg_shm->mobj;
+		}
+	}
+	cpu_spin_unlock_xrestore(&reg_shm_slist_lock, exceptions);
+	return NULL;
+}
+
+/*
+ * mobj_mapped_shm implementation.
+ * Used for registered SHM that is mapped to OP-TEE virtual address space
+ */
+
+struct mobj_mapped_shm {
+	struct mobj mobj;
+	tee_mm_entry_t *mm_entry;
+	struct mobj *reg_shm;
+	bool owns_reg_shm;
```

Comment:
is this `owns_reg_shm` necessary ? a non-null `reg_shm` already indicates a registered reg_shm mobj is realted to it.